File: web/src/ProviderService.ts
import ServiceBase from "./ServiceBase";
import { ProviderClient } from "./proto/ProviderServiceServiceClientPb";
import {
  InvitationStatusRequest,
  InvitationStatusResponse,
  InviteRequest,
  InviteResponse,
} from "./proto/ProviderService_pb";

export * from "grpc-web";
export class TrinsicProviderService extends ServiceBase {
  // channel: Channel;
  client: ProviderClient;

  constructor(serviceAddress: string = "https://localhost:5000") {
    super();

    // let credentials = ChannelCredentials.createInsecure();
    // let channel = new Channel(serviceAddress, credentials, {});
    // this.channel = channel;
    this.client = new ProviderClient(serviceAddress, {}, {});
  }

  // setChannel(channel: Channel) {
  //   this.channel = channel;
  //   this.client = new ProviderClient(
  //     channel.getTarget(),
  //     ChannelCredentials.createInsecure()
  //   );
  // }

  public inviteParticipant(request: InviteRequest): Promise<InviteResponse> {
    return new Promise((resolve, reject) => {
      this.client.invite(request, this.getMetadata(), (error, response) => {
        if (error) {
          reject(error);
        }
        return resolve(response);
      });
    });
  }

  public invitationStatus(request: InvitationStatusRequest): Promise<InvitationStatusResponse> {
    return new Promise((resolve, reject) => {
      this.client.invitationStatus(request, this.getMetadata(), (error, response) => {
        if (error) {
          reject(error);
        }
        return resolve(response);
      });
    });
  }
}

File: dotnet/Source/Trinsic/ProviderService.cs
﻿using System;
using System.Threading.Tasks;
using Grpc.Net.Client;
using Trinsic.Services;

namespace Trinsic
{
    public class ProviderService : ServiceBase
    {
        public ProviderService(string serviceAddress = "http://localhost:5000")
            : this(ServiceBase.CreateChannelIfNeeded(serviceAddress))
        {
        }

        public ProviderService(GrpcChannel channel)
        {
            // We must store a reference to the channel, otherwise it gets collected
            Channel = channel;
            ProviderClient = new Provider.ProviderClient(Channel);
        }

        public GrpcChannel Channel { get; }
        public Provider.ProviderClient ProviderClient { get; }

        /// <summary>
        /// Initates the participant onboarding flow using the input contact method
        /// </summary>
        /// <param name="request"></param>
        /// <returns></returns>
        public async Task<InviteResponse> InviteParticipant(InviteRequest request)
        {
            if (request.ContactMethodCase == InviteRequest.ContactMethodOneofCase.None)
            {
                throw new Exception("Contact method must be set.");
            }

            try
            {
                var response = await ProviderClient.InviteAsync(request, GetMetadata());
                return response;
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
        }

        /// <summary>
        /// Gets the status of the onboarding invitation for a given invitation ID
        /// </summary>
        /// <param name="request"></param>
        /// <returns></returns>
        public async Task<InvitationStatusResponse> InvitationStatus(InvitationStatusRequest request)
        {
            if (string.IsNullOrWhiteSpace(request.InvitationId))
            {
                throw new Exception("Onboarding reference ID must be set.");
            }

            try
            {
                var response = await ProviderClient.InvitationStatusAsync(request, GetMetadata());
                return response;
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
        }
    }
}
